File: C-sharp/Labwork 5/MainFlow/Validator.cs
using System.Text.RegularExpressions;

namespace Labwork_5.MainFlow
{
    public class Validator
    {
        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void ValidateAge(int age)
        {
            if (age <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(age), 
                    "The age of birthday celebrant musn't be less than 1");
            }
        }

        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void ValidateMeetingsCount(int meetingsCount)
        {
            if (meetingsCount < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(meetingsCount),
                    "The count of meetings per day can't be less than 0");
            }

            int maxMeetingsCount = 126;

            if (meetingsCount > maxMeetingsCount)
            {
                throw new ArgumentOutOfRangeException(nameof(meetingsCount),
                    "The count of meetings per day can't be "
                    +"more than 126 (at least 10 minutes per meeting and 3 hours for birthday)");
            }
        }

        /// <exception cref="ArgumentException"></exception>
        public static void ValidateName(string name)
        {
            if (!Regex.IsMatch(name, @"^[\p{L}\p{M}' \.\-]+$"))
            {
                throw new ArgumentException("The name is invalid");
            }
        }

        /// <exception cref="ArgumentException"></exception>
        public static void ValidatePlace(string place)
        {
            if (place == string.Empty)
            {
                throw new ArgumentException("The name of place shouldn't be empty");
            }
        }

        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void ValidateMeetingTime(DateTime meetingDateTime)
        {
            TimeOnly lastAvailableMeetingTime = new TimeOnly(20,50);

            if (TimeOnly.FromDateTime(meetingDateTime) > lastAvailableMeetingTime)
            {
                throw new ArgumentOutOfRangeException(nameof(meetingDateTime),
                    "The time of event shouldn't be bigger than last available (20:50)");
            }
        }

        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void ValidateBirthdayTime(DateTime birthdayDateTime)
        {
            TimeOnly mandatoryForBirthdayStart = new TimeOnly(21,0);

            if (TimeOnly.FromDateTime(birthdayDateTime) > mandatoryForBirthdayStart)                
            {
                throw new ArgumentOutOfRangeException(nameof(birthdayDateTime),
                    "The mandatory for birthday "
                    + $"celebrating is at 21 o'clock. Curret time is {birthdayDateTime}");
            }
        }

        /// <exception cref="ArgumentOutOfRangeException"></exception>
        public static void ValidateEventTime(Event occasion)
        {
            TimeSpan minMeetingDuration = new TimeSpan(0,10,0);

            if (ActivityScheduler.GetActivitiesList().Any(activity => 
                {
                    TimeSpan timeBetweenEvents = Event.GetTimeBetweenEvents(occasion,activity);

                    if (timeBetweenEvents < minMeetingDuration)
                    {
                        return true;
                    }

                    return false;
                }))
            {
                throw new ArgumentOutOfRangeException(nameof(occasion),
                    "The time between events should be at least 10 minutes");
            }
        }
    }
}

File: Python/Labwork 5/validator.py
from datetime import *

class Validator:
    @staticmethod
    def validate_name(name):
        if name == "":
            raise ValueError("Name musn't be empty")

    @staticmethod
    def validate_age(age):
        if age < 1:
            raise ValueError("The entered age musn't be less than 1")

    @staticmethod
    def validate_meetings_count(meetings_count):
        if meetings_count < 0:
            raise ValueError("The count of meetings per day can't be less than 0")

        max_meetings_count = 126

        if meetings_count > max_meetings_count:
            raise ValueError("The count of meetings per day can't be "
                + "more than 126 (at least 10 minutes per meeting and 3 hours for birthday)")

    @staticmethod
    def validate_place(place):
        if str(place) == "":
            raise ValueError("The name of place shouldn't be empty")

    @staticmethod
    def validate_meeting_time(meeting_date_time):
        last_available_meeting_time = time(20, 50)
        
        if meeting_date_time.time() > last_available_meeting_time:
            raise ValueError("The time of event shouldn't be bigger than last available (20:50)")

    @staticmethod
    def validate_birthday_time(birthday_date_time):
        mandatory_for_birthday_start = time(21, 0, 0)

        if birthday_date_time.time() > mandatory_for_birthday_start:
            raise ValueError("The mandatory for birthday celebrating is at 21 o'clock."
                + f" Current time is {birthday_date_time}")
